README.txt
PREREQUISITES
---
Lua 5.3: https://www.lua.org/download.html
Pd 0.51-0: http://msp.ucsd.edu/software.html

On MacOS make sure that Pd-0.51-0 is in the Application folder.
Check that Lua is properly installed running the following command from terminal

> lua -v

This should return something that looks like this:
> Lua 5.3.5  Copyright (C) 1994-2018 Lua.org, PUC-Rio



HOW TO USE
---
Launch the terminal and run the following commands:

> cd 'folder/containing/this/README/file'
> lua main.lua "platform"

where "platform" is:
- linux
- macos
- win

It is also possible to pass flags to Pd (ie -gui, -noaudio, etc. - not all are supported).



SUITE AND TESTS
---
A suite is a text file called "suite.txt" that contains a list of all the tests to run.
Each field (aka test) consists of the test name without extension (ie mytest_t) - see next section about tests for more info about the different file extensions.

Tests are pairs of Pd patches and text files sharing the same name but different extension (.pd and .txt).
For example "mytest_t" would be composed by the following files, placed in the /tests folder:
- mytest_t.pd - actual Pd patch to test
- mytest_t.txt - text file containing the expected results

The text file (ie mytest_t.txt) is a list of expected messages received from the correspondent Pd patch (ie mytest_t.pd).
It has 2 mandatory fields:
- teststart
- testend

All the other fields must be placed in between these 2 (see text files in /tests for some example).

The order of the fields in the .txt file should match the order of execution in the .pd file.
That means if the Pd patch sends the following messages:

1. teststart
2. append
3. float 23
4. banana
5. testend

the correspondent test .txt file should be:

```
teststart
append
float 23
banana
testend
```

to be noticed that the Pd test patch must always start the test sending "teststart" and finish it sending "testend". 


Both in "suite.txt" and the test .txt files, comments are allowed and start with "--" (ie "-- This is a comment!").

In the Pd patch (ie mytest_t.pd) results are passed to Lua via [stdout -cr].





INFO
---
All the test patches will automatically quit Pd after few seconds, if you want to edit it you can do one of the following:

Method 1
  - open the patch in a text editor
  - search for "EXIT" and turn it into "/EXIT" (or whatever else that is not "EXIT")
  - don't forget to set it back to "EXIT" once the changes have been completed

Method 2
  - launch the patch from terminal using the following command: 
    > pd -send "autoexitoff 1;" mytest_t.pd
    where "mytest_t.pd" is the name of the patch to edit



CHANGELOG.md
	2020-06-22
	---
	#ADDED
	- pdtest.lua - added check_test() to verify that test files are in the /tests folder

	#CHANGED
	- pdtest.lua - run() now checks that there is something in the "expected" results
	- pdtest.lua - check_flags() now check for argument passed


	2020-06-15
	---
	#ADDED

	#CHANGED
	- main.lua - added assert when opening suite.txt
	- pdtest.lua - get_fields() return false if no file has been opened - in case txt test file is not found test is skipped - nothing about .pd files is yet in place

	
	2020-06-10
	---
	#ADDED
	- pdtest.lua - run() now grabs Pd version and add it to the log
	- pdtest.lua - possibility to redirect stderr to stdout
	- float-int_t test

	#CHANGED
	- pdtest.lua - run() started code refactoring 
	- pdtest.lua - pd_version() now gets "platform" argument
	- logs - improved log layout
	- /tests - fixed pdcontrol_t.pd not closing and float-int_t.pd always closing regardless "autoexitoff" 
	- improved logs formatting
	- README.txt / INFO - fixed Method to open "autoclosing" test patches
	- main.lua - flags are not printed on the console anymore since the check happens in pdtest.lua
	- pdtest.lua - fixed issue where only one flag was considered

	2020-06-09
	---
	#ADDED
	- added possibility to pass flags (not all of them) to Pd
	- pdtest.lua - generate basic log.txt
	- added /logs folder
	- [list] test
	- main.lua - look for arg[1] before running any test
	
	#CHANGED
	- pdtest.lua - get_fields() now skips empty strings ("")
	- pdtest.lua - refactored "flag checker"
	- logs - now capture date, platform and flags
	- main.lua - now passes table containing tests to pdtest.lua
	- pdtest.lua - now prints expected vs actual when test fails
	- added failed_t - failing test
	- README.txt - added more info about the test architecture
	- pd test patches - placed all abstractions in /modules
	- pdtest.lua - now receives an arg instead of using arg[1]

	2020-06-08
	---	
	# ADDED
	- added EXIT subpatch to be used in all test patches	
	- /test folder containing few test examples
	- gitignore (just for emacs for now)
	- code ported

	# CHANGED
	- README.txt - added info about SUITE and TESTS	
	- pdtest.lua contains run() that actually runs the tests
	- utils.lua is now pdtest.lua


pdtest.lua
local pdtest = {}

local function parse_version(v)
   -- "pd -version" should return something like:
   -- Pd-0.49.1 ("") compiled 23:45:54 Jul 23 2019

   if v == nil then return "" end

   -- look for the apostrophe in the expected string - we can't look for "("
   local v_end = string.find(v, '"')
   if type(v_end) == "number" then
      v_end = v_end - 2
      if v_end > 0 then
	 return string.sub(v, 1, v_end)
      else
	 return ""
      end
   else
      return ""
   end
   
end


local function check_platform(platform)   

   local cmdlinux = "pd "
   local cmdmac = "/Applications/Pd-0.51-0.app/Contents/Resources/bin/pd "
      
   local pd_cmd = ""  
   if platform == "linux" then
      pd_cmd = cmdlinux 
   elseif platform == "macos" then
      pd_cmd = cmdmac
   elseif platform == "win" then
      print("well, we're not equipped for this yet!")
      os.exit()
   else
      print("First argument must be OS type:\nie: linux, macos, win")
      os.exit()
   end

   return pd_cmd
end



local function check_flags(flags)

   flags = flags or {}
   
   -- check flags to pass to Pd 
   local pdflags = {"-nogui", "-gui", "-nomidi", "-stderr", "-noaudio", "-r"} 
   local flags_str, redirect_stderr = "", ""

   if #flags == 0 then
      flags_str = "-nogui -stderr "

      -- redirect stderr to stdout
      redirect_stderr = "2>&1"

   else
      for k, v in ipairs(flags) do
	 for key, val in ipairs(pdflags) do
	    if v == val then
	       flags_str = flags_str .. " " ..  v .. " "

	       -- redirect stderr to stdout
	       if v == "-stderr" then redirect_stderr = "2>&1" end 
	    end
	 end
      end
      if flags_str == "" then flags_str = "-nogui -stderr " end
   end

   return flags_str, redirect_stderr
end




function pdtest.run(tests, platform, flags)
      
   local pd_cmd = check_platform(platform)
   
   flags = flags or {}

   local flags_str, redirect_stderr = check_flags(flags)
   
   
   local pd
   local timestamp = os.date("%Y%m%d_%H_%M_%S")
   local version = pdtest.pd_version(pd_cmd)
   local log = io.open("./logs/log_" .. timestamp .. ".txt", "w")   
   log:write("DATE: " .. os.date("%Y-%m-%d - %H:%M:%S") .. "\nPLATFORM: " .. platform .. "\nVERSION: " .. version .. "\nFLAGS: " .. flags_str .. "\n\n")


   
   -------------------
   -- run the tests --
   -------------------

   for k, testname in ipairs(tests) do
      
      local patchname = "./tests/" .. testname .. ".pd"
      local actual_cmd = pd_cmd .. flags_str .. patchname .. " " .. redirect_stderr
      
      log:write(string.format("\n---\nTEST: %s ", testname))
      io.write(string.format("\n---\n%s\n---\n", testname))

      
      -- check that testfiles actually exists
      local check_filename = pdtest.check_test(testname)
      if check_filename == false then
	 log:write("SKIPPED\n")
	 io.write("\nSKIPPED\n")
	 goto continue
      end

      
      
      -- launch Pd
      -- io.popen() is system depedent
      -- an alternative would be os.execute("cmd > ./log.txt") and then check the log
      pd = assert(io.popen(actual_cmd))


      -- expected msg from Pd
      local expected = pdtest.get_fields("./tests/" .. testname .. ".txt")
      if #expected == 0 then
	 log:write("SKIPPED\n")
	 io.write("\nSKIPPED\n")
	 goto continue
      end

      

      -- where we collect what Pd actually sends us
      local actual = {}
      local pderr = {}


      

      ---------------------
      -- collect from Pd --
      ---------------------

      local line = ""
      local start = false
      local c = 0
      while line ~= nil do
	 if line == "teststart" then
	    start = true
	 end
	 if start then
	    c = c + 1
	    actual[c] = line
	 end

	 -- intercept stderr
	 if string.find(line, "error:") ~= nil then
	    table.insert(pderr, line)
	 end
	 
	 line = pd:read()
      end


      
      -----------------------------
      -- verify things collected --
      -----------------------------

      local failed = {}

      -- check that we actually got something from Pd
      if #actual > 0 then
	 for k, v in ipairs(expected) do
	    local result = actual[k] == v
	    
	    -- if "actual" is different from "expected" capture both
	    if result == false then table.insert(failed, {v, actual[k]}) end
	 end
	 
	 if #failed > 0 then
	    io.write("\nFAILED\n")
	    log:write("FAILED\n")
	    for i=1, #failed do
	       io.write(string.format("- expected: %s | actual: %s\n", failed[i][1], failed[i][2]))
	       log:write(string.format("- expected: %s | actual: %s\n", failed[i][1], failed[i][2]))
	    end
	    io.write("\n")
	 else
	    io.write("\nTEST PASSED\n\n")
	    log:write("PASSED\n")
	 end

	 -- if for whatever reason we haven't got anything from Pd
      else
	 io.write("\nTEST FAILED - getting nothing from Pd\n")
	 log:write("FAILED - getting nothing from Pd\n")
      end

      -- log Pd stderr first
      if #pderr > 0 then
	 log:write("\nPd errors\n")
	 for k, v in ipairs(pderr) do
	    io.write("Pd " .. v, "\n")
	    log:write(v .. "\n")
	 end
      end
      ::continue::
   end

   log:close()
   pd:close()
end



function pdtest.check_test(filename)
   -- check if both filename.txt and filename.pd exist
   -- if any of the 2 is not there return false otherwise true
   
   if filename == nil then
      return false
   end

   local textfile = io.open("./tests/" .. filename .. ".txt")
   local pdfile = io.open("./tests/" .. filename .. ".pd")

   -- if any of the file is not in the folder
   if textfile == nil or pdfile == nil then
      return false
   end


   -- if both the files are in the folder
   textfile:close()
   pdfile:close()
   
   return true
   
end



function pdtest.get_fields(filename)
   -- get test expected results from test .txt file

   local file
   
   if filename == nil then
      return false
   else
      file = io.open(filename)
   end
   if file == nil then return false end

   
   local expected = {}
   local line = "--"
   while line ~= nil do
      -- skip comments
      if string.sub(line, 1, 2) ~= "--" and line ~= "" then
	 table.insert(expected, line)
      end
      
      if line == "testend" then break end

      -- we skip newline and return
      line = file:read("l")
   end

   return expected
end



function pdtest.pd_version(cmd)
   local pd = io.popen(cmd .. "-version 2>&1")
   local version = ""
   
   local line = ""
   while line ~= nil do
      if line ~= "" then
	 version = parse_version(line)
      end
      line = pd:read("l")
   end

   pd:close()

   return version
end




return pdtest
